// @lwc/template-compiler/dist/commonjs/shared/parse5.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createEmptyElementLocation = exports.getTemplateContent = exports.isTextNode = exports.isCommentNode = exports.isElementNode = void 0;
function isElementNode(node) {
    return 'tagName' in node;
}
exports.isElementNode = isElementNode;
function isCommentNode(node) {
    return node.nodeName === '#comment';
}
exports.isCommentNode = isCommentNode;
function isTextNode(node) {
    return node.nodeName === '#text';
}
exports.isTextNode = isTextNode;
function getTemplateContent(templateElement) {
    return templateElement.content;
}
exports.getTemplateContent = getTemplateContent;
// Creates a parse5.ElementLocation where all values are set to 0.
function createEmptyElementLocation() {
    const startTag = createEmptyStartTagLocation();
    const endTag = createEmptyLocation();
    const elementLocation = createEmptyStartTagLocation();
    return { ...elementLocation, startTag, endTag };
}
exports.createEmptyElementLocation = createEmptyElementLocation;
function createEmptyStartTagLocation() {
    return { attrs: {}, ...createEmptyLocation() };
}
function createEmptyLocation() {
    return { startCol: 0, startOffset: 0, startLine: 0, endCol: 0, endOffset: 0, endLine: 0 };
}
